### PR TITLE
Basic JSX support

### PR DESCRIPTION
**Pull request by jridgewell**

### CODE REVIEW

**Comment by jridgewell** [comment]
Haven't gotten text to work yet, any ideas?

File: .jshintrc [hunk 1]
```diff
@@ -0,0 +1,62 @@
+{
```
Comment (thejameskyle) [inline .jshintrc]: jshint? what year is this :stuck_out_tongue:

Comment (jridgewell) [inline .jshintrc]: It's been on my system for so long it's hard not to just use it... Let me clean up the PR.

**Comment by thejameskyle** [comment]
I don't have much time to look at this now, but you can see how this works in Babel's [react transformer](https://github.com/babel/babel/blob/master/packages/babel/src/transformation/helpers/react.js).

I wouldn't use a `Literal` visitor and check the parent, You're going to visit way too many nodes.

**Comment by jridgewell** [comment]
Well, that worked better than I thought. Any ideas to clean up, when you get the chance?

File: test/fixtures/key/expected.js [hunk 2]
```diff
@@ -1,4 +1,10 @@
 "use strict";
 
-elementOpen("div", "1");
+elementOpen("div");
+elementOpen("div", "1", ["key", "1"])
+elementClose("div")
```
Comment (thejameskyle) [inline test/fixtures/key/expected.js]: These need to be expression statements, `elementClose(...)elementOpen(...)` will break.

File: package.json [hunk 3]
```diff
@@ -12,8 +12,7 @@
   "scripts": {
     "build": "babel-plugin build",
     "push": "babel-plugin publish",
-    "test": "babel-plugin test",
-    "test-mocha": "mocha --compilers js:babel/register"
+    "test": "mocha --compilers js:babel/register"
```
Comment (thejameskyle) [inline package.json]: Asked about this here https://github.com/babel/babel/issues/2029

File: src/index.js [hunk 4]
```diff
@@ -1,29 +1,43 @@
...
-        if (attr.name.name === "key") {
-          key = attr.value.value;
-        } else {
-          attrs.push(
-            t.literal(attr.name.name),
-            t.literal(attr.value.value)
-          );
+      parent.children = buildChildren(parent);
+
+      for (let { name: { name: attr }, value } of node.attributes) {
```
Comment (thejameskyle) [inline src/index.js]: Not a fan of the deeply nested destructuring, this seems more confusing.

File: test/fixtures/key/expected.js (hunk 2 above)
Comment (jridgewell) [inline test/fixtures/key/expected.js]: Excellent, that's exactly what I needed to know.

Now, any way to preserve the indentation? It's minor, but makes reading the output much easier.

File: src/index.js (hunk 4 above)
Comment (jridgewell) [inline src/index.js]: I'll clean that up.

File: test/fixtures/key/expected.js (hunk 2 above)
Comment (thejameskyle) [inline test/fixtures/key/expected.js]: You'd have to hook into the code generation. I don't think that's exposed at all. cc @sebmck

File: test/fixtures/spread-attributes/expected.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+"use strict";
+
+elementOpenStart("div", "test", ["class", "test", "key", "test"]);
```
Comment (thejameskyle) [inline test/fixtures/spread-attributes/expected.js]: is `"key", "test"]` supposed to be put here?

Comment (jridgewell) [inline test/fixtures/spread-attributes/expected.js]: For support, I think so. Idom will [initialize stored data](https://github.com/google/incremental-dom/blob/master/src/node_data.js#L105-L107) (if it doesn't exist) with a key from the `key` attribute.

**Comment by jridgewell** [comment]
Anything else blocking this?

**Comment by thejameskyle** [comment]
Can you add some code comments?

**Comment by jridgewell** [comment]
Commented.

File: src/index.js [hunk 6]
```diff
@@ -1,70 +1,71 @@
...
-          key = attr.value.value;
-        } else {
-          attrs.push(
-            t.literal(attr.name.name),
-            t.literal(attr.value.value)
-          );
-        }
+      // Only push arguments if they're needed
+      if (key || statics || attrs) {
+        args.push(key ? key : t.literal(null));
```
Comment (sebmck) [inline src/index.js]: Could just be `key || t.literal(null)`.

**Comment by thejameskyle** [comment]
Awesome work @jridgewell

**Comment by jridgewell** [comment]
Whoot! :smile:

**Pull request merged**